Searchbar: Launch on enter / done

Sir-Photch opened the pull request:
Closes #53.

Currently, it will only launch the first app that is found. I suppose it would make sense to add other searchables as well.

Things I'd still want to add:
- [x] Pass search query to first quickaction, if no other searchable is found

MM2-0 commented on app/ui/src/main/java/de/mm20/launcher2/ui/launcher/search/SearchVM.kt:
```diff
@@ -69,7 +73,23 @@ class SearchVM : ViewModel(), KoinComponent {
+
+        suspend {
+            searchJob?.join()
+        }
```
What do you expect this to do?

MM2-0 commented on app/ui/src/main/java/de/mm20/launcher2/ui/launcher/search/SearchVM.kt:
```diff
@@ -65,11 +73,40 @@ class SearchVM : ViewModel(), KoinComponent {
+            contactResults,
+            calendarResults,
+            fileResults
+        ).map { it.value ?: emptyList() }
```
This can be optimized using `.firstNotNullOfOrNull { it.value?.firstOrNull() }`

MM2-0 commented:
Looks good so far. One thing we should add is some sort of visual indication to let users know beforehand which item will be launched. But I can add that later.

Sir-Photch commented:
> One thing we should add is some sort of visual indication to let users know beforehand which item will be launched. But I can add that later.

Yes, I have also thought about this: [iOS is doing something similar](https://support.apple.com/en-us/HT201285) where they put a border around the best match if it is an app. I think something like this would look nice as well. For the case when no match is found, you could default to highlighting the first quickaction.

Sir-Photch commented:
I've added the back-end necessary for highlighting, and a rough prototype how I imagine it to look like for the AppResults-GridView:

<details>
<summary>Screenshot</summary>

![searchbar](https://user-images.githubusercontent.com/47949835/219867691-13e4c791-80af-4401-885d-980fb8083cdc.png)

My idea will obviously only work for non-square icon shapes right now. Also, the circle is cut off.
Though judging by your changes for #253, I am sure you have better Ideas how to integrate this into the launchers appearance :)

</details>

Things missing:

- [ ] Highlighting for `ListItem`
- [ ] Default highlighting of first quick-action

Sir-Photch commented:
For squared icons with rounded corners it looks nice, but for square icons it doesn't fit as well:

<details>
<summary>Screenshot</summary>

![highlight](https://user-images.githubusercontent.com/47949835/219880607-6fe03c20-4caa-482e-a227-d47ac9e821f8.png)

</details>

Maybe it'd be possible to choose this outline-box based on what the current icon-shape is?

MM2-0 commented:
> Maybe it'd be possible to choose this outline-box based on what the current icon-shape is?

It is, you can use `LocalIconShape.current` to access it. I was considering it but liked the rounded rect more. But you have a good point, and with that in mind, adapting the selection box to the icon shape is objectively the better solution.

<details>

![photo_2023-02-18_18-51-42](https://user-images.githubusercontent.com/15646950/219880836-572a0973-f220-4af9-9e49-33abee1b22ad.jpg)

![photo_2023-02-18_18-51-43](https://user-images.githubusercontent.com/15646950/219880838-3d5fb24c-3890-45f6-b156-5d5d569e6ac1.jpg)

![photo_2023-02-18_18-51-44 (2)](https://user-images.githubusercontent.com/15646950/219880839-91882dc9-3a6a-4ec4-ba5f-385b0286a0ce.jpg)

![photo_2023-02-18_18-51-44](https://user-images.githubusercontent.com/15646950/219880840-0c28f83f-5b72-4ad7-be04-3c3be55130fe.jpg)


</details>

Sir-Photch commented:
I've changed it to the icon shape. One weird thing I have noticed with the reuleaux-triangle:

<details>
<summary>Screenshot</summary>

![reuleaux](https://user-images.githubusercontent.com/47949835/219901558-9f42da82-44dd-48f7-8187-087860ce0682.png)

</details>

Otherwise, highlighting looks fine.
Also, I have adjusted the border-size to `Dp.Hairline`, since otherwise I find it too pronounced.
